### ISSUE TITLE
define custom gene list

### ISSUE THREAD

**Issue by Mohammedhusen** [cBioPortal/cbioportal#7524]
Hi Team,

Is it possible to define our own gene list instead of entering every time in "Enter Genes:
" dropdown?

**Comment by jjgao** [comment]
This would be a nice feature.

@kalletlak I am wondering if it's possible to save custom gene list to session service?

**Comment by kalletlak** [comment]
@jjgao it's possible.
- [ ] Identify places where we need to have this feature
- [ ] Define the model that needs to be saved

**Comment by jjgao** [comment]
maybe we can add a button "Save this gene set" when it is "User-defined list", which will bring out a modal dialog to input the list name with the genes prefilled. Once it's save, it'll be save to the users profile and be listed as the first gene list in the drop down (right after User-defined list).

![image](https://user-images.githubusercontent.com/840895/83191786-9a5cc900-a102-11ea-80da-05f5e59ef404.png)

**Comment by jjgao** [comment]
@cBioPortal/product thoughts?

**Comment by schultzn** [comment]
Great idea! Just need a good UI to allow users to name / rename / modify / remove the gene lists.

**Comment by jjgao** [comment]
Maybe something like this for rename and delete? Save/name is above.

![image](https://user-images.githubusercontent.com/840895/83281387-08a69780-a1a6-11ea-8f9a-5b15be739fa0.png)

**Comment by schultzn** [comment]
We use the trashcan and pencil icon in other places - we should probably be consistent.

**Comment by jjgao** [comment]
>We use the trashcan and pencil icon in other places - we should probably be consistent.

Sure, maybe rename can be an edit icon and save can be a save icon... They will be less visible but that's probably good.

**Comment by jjgao** [comment]
@jtquach1 I am wondering if you may be able to help with this one now that you have some experience with session service? @Luke-Sikina @kalletlak

**Comment by alisman** [comment]
@cBioPortal/product @dippindots we use react-select for this dropdown, so it should be pretty easy to add the edit buttons in line with the menu options.   I.e. you open the menu, you see custom lists and each one has edit options next to it.  

Also, I think that the button could be a sibling of the validation message like so:

![image](https://user-images.githubusercontent.com/186521/125834828-37e59608-4e1f-4e89-9f89-7eb3313a9e0f.png)

**Comment by dippindots** [comment]
Proposed solution:

**Save user defined list into user profile**
Click `Save custom gene list` to open a tooltip to save gene list into user profile:
![image](https://user-images.githubusercontent.com/15748980/127509916-351ce416-88c4-49bd-beb5-5d4e2b7b758a.png)

Tooltip:
![save gene](https://user-images.githubusercontent.com/15748980/127517836-cd25e259-d337-4a9c-8529-d04d8061efe2.png)

**Edit Or Delete a saved list**
Click icon on the right side of the gene list name, and open a tooltip.
![image](https://user-images.githubusercontent.com/15748980/127509823-a4898130-a9a8-451e-8021-f201396817e4.png)

**Tooltip 1: Edit gene list**
![edit gene](https://user-images.githubusercontent.com/15748980/127519165-da286546-3cbc-4b7f-b2c7-52506441acf9.png)

**Tooltip 2: Delete gene list**
![delete gene](https://user-images.githubusercontent.com/15748980/127520408-8148975d-4e91-4169-b447-f4f3938dcc1a.png)

**Comment by alisman** [comment]
@dippindots doesn't need to be in a tooltip.  just do a dialog box.  

as i proposed above, if it's all the same, i think it would be better if you you included the pencil icon and trash button at the option level inside the menu so that you don't have to choose the item in order to see these options.    i would think this would be just as easy to accomplish using react-select.

**Comment by dippindots** [comment]
> @dippindots doesn't need to be in a tooltip. just do a dialog box.
Ok, if everyone agree to have a modal rather than tooltip, we will make a modal then.
CC: @cBioPortal/product @jjgao 
> as i proposed above, if it's all the same, i think it would be better if you you included the pencil icon and trash button at the option level inside the menu so that you don't have to choose the item in order to see these options. i would think this would be just as easy to accomplish using react-select.
Maybe I didn't make it clear, my proposal is based on yours, and I want to add those two icons in the option of react-select. You should be able to modify any options in the dropdown list or after you selected one option.

**Comment by tmazor** [comment]
The proposal above looks good to me! I assume this will only work for logged in users? Will we also display the button for not logged in users, but disabled with a prompt to log in (or some other solution along those lines)?

**Comment by kalletlak** [comment]
@dippindots this looks good. I have couple of questions to product team

1. Is this list name used anywhere? Also specific reason to allow editing?
2. Are these list shareable? if it is then I think it should treated similar to group or custom chart in study page.
3. Is it study specific? or common for all studies?
4. For logged in users - If person A create custom list and query it and then share the results page url to person B - will that custom list be added to person B?

**Comment by dippindots** [comment]
Thank you so much @kalletlak for your comments! @cBioPortal/product Could you help answer these questions we have?

> @dippindots this looks good. I have couple of questions to product team
> 
> 1. Is this list name used anywhere? Also specific reason to allow editing?
> 2. Are these list shareable? if it is then I think it should treated similar to group or custom chart in study page.
> 3. Is it study specific? or common for all studies?
> 4. For logged in users - If person A create custom list and query it and then share the results page url to person B - will that custom list be added to person B?

**Comment by alisman** [comment]
@kalletlak agreed.  seems like a feature that we would want to share in the same way as we do groups.  and that would then mean that they need to be immutable, correct?

**Comment by kalletlak** [comment]
@alisman yes

**Comment by jjgao** [comment]
> 1. Is this list name used anywhere? Also specific reason to allow editing?

Nop, only here. Doesn't have to be editable - maybe a deletion option is enough.

> 2. Are these list shareable? if it is then I think it should treated similar to group or custom chart in study page.

I don't think that's necessary.

> 3. Is it study specific? or common for all studies?

Not study specific.

4. For logged in users - If person A create custom list and query it and then share the results page url to person B - will that custom list be added to person B?

Nop.

Overall, I think we should simplify this feature. It can live in a single user's space without sharing functionality. It doesn't need to get into the URL.

**Comment by kalletlak** [comment]
Sure. Looks like we are just sending gene list to result page, list id or name is never used.
